_posts/2017-09-08-scalafix-v0.5.md

```diff
@@ -0,0 +1,191 @@
+---
+category: blog
+by: Ólafur Páll Geirsson
+title: "Announcing Scalafix v0.5"
+---
+
+I am excited to announce the release of [Scalafix v0.5.0][].
+This release introduces new features in addition to several bug fixes and
+integration improvements.
+
+If you are not familiar with Scalafix, Scalafix is a rewrite and linting tool
+for Scala.
+The project is developed at the [Scala Center][sc] with the mission to help
+automate migration between different Scala compiler and library versions.
+
+> This is the third post on Scalafix and scalameta.
+> You might be interested in reading the previous post
+> [here](http://www.scala-lang.org/blog/2017/02/27/Scalafix-v0.3.html).
+
+
+## Linting
+
+Scalafix v0.5 supports linters.
+Before v0.5, Scalafix rules could only "fix" code by providing rewrite instructions.
+Now, it's possible to implement Scalafix rules that report lint messages instead
+of rewrite instructions.
+Each category of lint messages is attached to a default severity level
+(info/warn/error), which users can override via configuration.
+
+The first Scalafix linter is [`NoInfer`][].
+`NoInfer` reports errors when the Scala compiler infers one of the following types:
+`Any`, `AnyVal`, `Product` or `Serializable`.
+It's usually a bad sign when the compiler infers such a generic type.
+In those cases, it's better to explicitly type annotate the expected type.
+
+
+### Building a custom linter
+It's simple to implement custom Scalafix linters.
+To demonstrate this, let's implement a `NoDefaultParams` linter that reports an error if
+default parameter values are used.
+To begin with, we motivate this linter rule with a bit of documentation.
+
+<script src="https://gist.github.com/olafurpg/38699f591329c3bcff52b7412dcb696b.js"></script>
```

Comment:
Is there any way that we can keep the code examples in the markdown file?

Reply:
The reason I keep the listings in gists is to demonstrate that a linter "project" can be released with minimal overhead.

- write docs in gist, see https://gist.github.com/olafurpg/38699f591329c3bcff52b7412dcb696b
- share linter implementation via gist, https://gist.github.com/olafurpg/7b022c7120156716def282028b091e8b
- users can run linter from gist, as shown with `> sbt https:/....` below

I'll inline the implementation and test, but the docs snippet is markdown so it would look weird inlined.

Reply:
Why would it look weird inlined? You just do:

```
## Header 2

This is some _text_!.
```

Reply:
Further, this point that you're making "The reason I keep the listings in gists is to demonstrate that a linter "project" can be released with minimal overhead" was certainly not obvious from the blog article. Did you intend for it to be?

Reply:
I've updated the "Building a custom linter" section to 

- be about `FinalCaseClass` (report on missing `final` on case classes) instead of `NoDefaultParams`
- replace gists with inline code listings
- use the scalacenter/scalafix.g8 template
- add link to git repo project olafurpg/FinalCaseClass

I hope it helps clarify the "minimal overhead project".